### PR TITLE
Product Filters > Fix Performance issue and Fatal error on stores with a high volume of products

### PR DESCRIPTION
**Pull request by nefeline**
This PR fixes the issue initially reported on pdnLyh-3VR-p2 by removing the SQL queries used to manipulate the product counts for filters in particular cases. These queries were triggered for 100% of the stores using the Products (Beta) or the Product Collection blocks, independent of their usage of the Product Filters. Consequently, they negatively affected all stores' performance and triggered memory exhaustion errors for big e-commerces with thousands of products.

## What these queries were used for?

A practical use-case example is: when adding the Products (beta) block on any post or page, it is possible to edit a setting under "Advanced Filters" that allows you to handpick specific products for display instead of showing all of them by default:

Screenshot 2023-07-13 at 19 33 58

If any user decided to use the Filter by Attribute, Filter by Rating, or the Filter by Stock blocks with the filter counts enabled (they are disabled by default) alongside the Products (Beta) block that is just rendering a handful of products, these queries would manipulate the product counts to match the displayed products, and not all available products in the store right before rendering the block.

In other words, these queries were modifying the native behavior of the filter counts, and as a consequence, they were [not aligned with the design decisions](https://github.com/woocommerce/woocommerce-blocks/discussions/9023) made a few cycles ago. With the changes introduced here, all filter counts now provide the same functionality across the board, independent of their use cases.

## Discussion

There's an ongoing internal discussion, opened by @kmanijak, to gather feedback from the design team in case we wanna open an exception for this use case in particular and deviate from the design decision made previously. Depending on the outcome of that discussion, more changes will follow. (ref. per0F9-Rh-p2 ) 

Its also important to note that these queries are not necessary for maintaining the functionality of the filters or the filter counts; the [get_attribute_counts](https://github.com/woocommerce/woocommerce-blocks/blob/0d809ee287e52d3a348448349be5bf094bdb88ae/src/StoreApi/Utilities/ProductQueryFilters.php#L167) method is.

## User Facing Testing

1. Make sure you have products in your store with attributes and stock status (You can do so by simply importing the sample products from the core of Woo).
2. Create a new post.
3. Insert the Products (Beta) block.
4. Insert the Filter by Attribute, Filter by Stock and Filter by Price
5. Make sure all filters have the "Display product count" setting enabled (it is disabled by default).

Screenshot 2023-07-13 at 20 09 31
 
6. Save the post and go to the front-end
7. Make sure you can use all filters as expected and without any problems/errors: the product counts should match the number of existing products in the store with the filtered criteria, not just what is displayed on the page.
8. Repeat all steps from 2 to 7, but using the Products Collection block instead of Products (beta). Make sure all filters work as expected.

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

This update will positively impact performance as two heavy queries were removed from the codebase. Stores with a high volume of products should be able to use the Products Collection and Products (Beta) blocks without risking having memory exhaustion errors as a consequence.

### Changelog

> Improve performance and fix memory exhaustion errors that could be triggered for stores with a high volume of products using the Products (Beta) or Products Collection blocks.

### CODE REVIEW

**Review by kmanijak — APPROVED** [review]
Thanks for this PR 🙌 

I agree it brings consistency to how the filter works as agreed in the discussion [#9023](https://github.com/woocommerce/woocommerce-blocks/discussions/9023) and it's great it fixes the performance problem!

I'm approving, but before merging, could you please remove references to All Products block from the PR description as I don't think it's affected by this problem at all. It's Products (Beta) and Product Collection (which is still experimental and not available to use). Unless I'm missing something?

As a next step, let's keep discussing filters along with the custom-configured Products (Beta) block (but that's the next chapter!)

**Comment by imanish003** [comment]
Hi @nefeline 👋  

> it is possible to edit a setting under "Advanced Filters" that allows you to handpick specific products for display instead of showing all of them by default:

I think it’s applicable for all filters instead of just the *handpicked products* filter. If any filter changes the number of products that will be visible on frontend then this is applicable. For example:

- Stock status: If user set stock status as *“On backorder”* then we show the products which are on backorder
- Same with other filters like Product categories, product tags or keyword, and so on.



In short, if there is any filter that we apply in *Products beta* or *Product collection block*, which reduces the number of products to be visible on frontend then it will be applicable. 

I'm not entirely certain if we're on the same page regarding this, so I wanted to make sure I overcommunicate here to prevent any potential confusion 🙂

**Comment by nefeline** [comment]
Thanks for the review @kmanijak and @imanish003 !

Regarding:

> I'm approving, but before merging, could you please remove references to All Products block from the PR description as I don't think it's affected by this problem at all. It's Products (Beta) and Product Collection (which is still experimental and not available to use). Unless I'm missing something?

Yep, good call! I replaced all occurrences of All Products with Product Collection.

Can any of you please re-review this PR 🙏 ? I pushed some changes on [1b4131f](https://github.com/woocommerce/woocommerce-blocks/pull/10198/commits/1b4131f34d425fff9e7610541d8757c994e2b7db) as I noticed the filters were not working alongside the Product Collection block, now they do.

**Comment by nefeline** [comment]
> I think it’s applicable for all filters instead of just the handpicked products filter.

Yep! Handpicked products was just a use-case example.

> A practical use-case example is: when adding the Products (beta) block on any post or page[...]

**Review by kmanijak — APPROVED** [review]
> Can any of you please re-review this PR 🙏 ? I pushed some changes on [1b4131f](https://github.com/woocommerce/woocommerce-blocks/pull/10198/commits/1b4131f34d425fff9e7610541d8757c994e2b7db) as I noticed the filters were not working alongside the Product Collection block, now they do.

I double-checked Products and Product Collection ✅
